blog/sage-upgrading-a-smart-contract/sage-upgrading-a-smart-contract.md

```diff
@@ -0,0 +1,218 @@
+    address public implementation;
+
+    constructor(address _implementation) {
+        implementation = _implementation;
+    }
+
+    fallback() external payable {
+        address _impl = implementation;
+        assembly {
+            let ptr := mload(0x40)
+            calldatacopy(ptr, 0, calldatasize())
+            let result := delegatecall(gas(), _impl, ptr, calldatasize(), 0, 0)
+            let size := returndatasize()
+            returndatacopy(ptr, 0, size)
+
+            switch result
+            case 0 { revert(ptr, size) }
+            default { return(ptr, size) }
+        }
+    }
+}
+
+contract CounterFactory {
+    function createCounter() public returns (address) {
+        Counter counter = new Counter();
+        CounterProxy proxy = new CounterProxy(address(counter));
+        return address(proxy);
+    }
+}
+```
+
+In the code above, the "CounterFactory" contract includes a function called "createCounter," which generates a new instance of the "Counter" contract as well as a new instance of the "CounterProxy" contract. The "Counter" contract's address is passed to the "CounterProxy" contract's constructor. The address of the proxy contract is returned by the function.
+
+### Step4: Upgrade the Implementation Contract
+
+The next step is to upgrade the implementation contract to include a new function that resets the counter to zero.
+
+Counter.sol
+
+```solidity
+pragma solidity ^0.8.0;
+
+contract Counter {
+    uint256 public count;
+
+    function increment() public {
+        count++;
+    }
+
+    function reset() public {
+        count = 0;
+    }
+}
+```
+
+### Step5: Upgrade the Proxy Contract
+
+The final step is to modify the proxy contract so that it points to the new implementation contract. We have to deploy a new instance of the implementation contract and update the proxy contract's "implementation" variable to point to the new contract address.
+
+```solidity
+contract CounterFactory {
+    function createCounter() public returns (address) {
+        Counter counter = new Counter();
+        CounterProxy proxy = new CounterProxy(address(counter));
+        return address(proxy);
+    }
+
+    function upgradeCounter(address _proxy) public {
+        Counter counter = new Counter();
+        CounterProxy proxy = CounterProxy(_proxy);
+        proxy.implementation = address(counter);
+    }
+}
+```
+
+A new "upgradeCounter" function takes the address of an existing proxy contract and upgrades it to use the new implementation contract. The function creates a new instance of the implementation contract and then modifies the proxy contract's "implementation" variable to point to the new contract address.
+
+To upgrade the contract, we can use the address of the existing proxy contract to call the "upgradeCounter" function. This will update the proxy contract's implementation contract, allowing us to add new functionality to the contract without having to deploy a new proxy contract.
```

Comment:
Hey just write that where to store the files, like making a directory and then saving them all in one place or something like use remix.ide and make this file and steps like that which will help the beginners in a better way